homework_4_functions/main.js

```diff
@@ -0,0 +1,149 @@
+//* 1. add all arguments passed to a function
+
+const addThemAll = (...numbers) => numbers.reduce((acc, num) => acc + num, 0);
+
+console.log(addThemAll(2, 4)); // 6
+console.log(addThemAll(1, 2, 3, 4)); // 10
+console.log(addThemAll(5, 5, 10)); // 20
+
+//* 2. multiply with closure
+
+const multiply = (a) => (b) => a * b;
+
+console.log(multiply(5)(5)); // 25
+console.log(multiply(2)(-2)); // -4
+console.log(multiply(4)(3)); // 12
+
+//* 3. movie sorting
+
+const movies = [
+  {
+    movieName: 'The Thing',
+    releaseYear: 1982,
+    directedBy: 'Carpenter',
+    runningTimeInMinutes: 109,
+  },
+  {
+    movieName: 'Aliens',
+    releaseYear: 1986,
+    directedBy: 'Cameron',
+    runningTimeInMinutes: 137,
+  },
+  {
+    movieName: 'Men in Black',
+    releaseYear: 1997,
+    directedBy: 'Sonnenfeld',
+    runningTimeInMinutes: 98,
+  },
+  {
+    movieName: 'Predator',
+    releaseYear: 1987,
+    directedBy: 'McTiernan',
+    runningTimeInMinutes: 107,
+  },
+];
+
+const byProperty = (property, direction) => {
+  const sortAscending = (movie1, movie2) => {
+    if (movie1[property] < movie2[property]) {
+      return -1;
+    }
+    if (movie1[property] > movie2[property]) {
+      return 1;
+    }
+    return 0;
+  };
+
+  const sortDescending = (movie1, movie2) => {
+    if (movie1[property] > movie2[property]) {
+      return -1;
+    }
+    if (movie1[property] < movie2[property]) {
+      return 1;
+    }
+    return 0;
+  };
+
+  if (direction === '<') {
+    return sortDescending;
+  }
+
+  return sortAscending;
+};
+
+console.log(movies.sort(byProperty('releaseYear', '>'))); // виведе масив фільмів посортованих по року випуску, від старішого до новішого*
+console.log(movies.sort(byProperty('runningTimeInMinutes', '<'))); // виведе масив фільмів посортованих по їх тривалості, від найдовшого до найкоротшого*
+console.log(movies.sort(byProperty('movieName', '>'))); // виведе масив фільмів посортованих по назві, в алфавітному порядку*
+
+//* 4.1 detonatorTimer via setInterval()
+
+const detonatorTimer1 = (delay) => {
+  let counter = delay;
+  const intervalId = setInterval(() => {
+    if (counter) {
+      console.log(counter);
+      counter -= 1;
+    } else {
+      console.log('BOOM!');
+      clearInterval(intervalId);
+    }
+  }, 1000);
+};
+
+detonatorTimer1(3);
+// 3
+// 2
+// 1
+// BOOM!*
+
+//* 4.2 detonatorTimer via setTimeout()
+
+const detonatorTimer2 = (delay) => {
+  if (!delay) {
+    setTimeout(() => console.log('BOOM!'), 1000);
+    return;
+  }
+  setTimeout(() => {
```

Comment:
Looks line we have the `setTimeout` in the case with the `delay` and without them. Can we avoid this copy/paste?

Reply:
I put the recursion exit check inside the setTimeout() too. It seems like it is better now, but I am not 100% sure 🤔